### ISSUE TITLE
[red-knot] special-case returning NotImplemented

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#16661]
`NotImplemented` should always be returnable, no matter what return type annotation a function has.

Other type checkers seem to handle this by just treating `NotImplemented` as if it were `Any`, but [I don't think that's right](https://github.com/astral-sh/ruff/issues/16652).

We can easily handle this by treating `NotImplemented` as a `KnownInstance` and special-casing it in return type checking.

(Arguably this special case should only apply to certain methods, but that seems more complex and perhaps not worth it.)

**Comment by cake-monotone** [comment]
I found that there may be additional cases we should consider

```python
class CustomInt(int):
    def __add__(self, o) -> CustomInt:
        # the type of `add_result` will be inferred as `int`, but actually it will be `int | _NotImplementedType` at runtime.
        add_result = int.__add__(int(self), o)
        if add_result is NotImplemented:  # this condition will always be inferred as Literal[False]
            return NotImplemented

        return add_result * add_result
```

Treating `NotImplemented` as `Any` might be intended to cover cases like this.
If we treat `NotImplemented` as a `KnownInstance`, we might need to ensure that the result of binary operations is not always inferred as `Literal[True]` or `Literal[False]`. What do you think?

**Comment by carljm** [comment]
If you perform a binary operation that de-sugars to a dunder method call, `NotImplemented` is not a possible result of that operation, because the runtime handles `NotImplemented` internally to fall back to another behavior (e.g. calling a reverse-dunder method). So this issue only occurs when manually/explicitly calling a dunder method (which is generally discouraged), as in your example.

I don't think there's much we can do about this, because it's an inaccuracy that is baked in to the typeshed definitions of dunder methods and their return types.

It's also not clear to me how treating `NotImplemented` as `Any` (rather than special-casing it in return checking) would help with your example? I guess you are worried that we might wrongly mark the `return NotImplemented` branch as dead code?

Overall I think if this scenario can only be triggered with a manual call to a dunder method, then it is OK if that results in some sub-optimal behavior. It seems like a rare edge case.

**Comment by cake-monotone** [comment]
You're right — I was a bit concerned about the dead code branch. But if this case isn’t something we need to specifically account for, I’ll go ahead and implement it as I initially intended. Would you mind assigning the issue to me?
